### ISSUE TITLE
mat: SymDense symmetric "shearing" transformation?

### ISSUE THREAD

**Issue by decibelcooper** [gonum/gonum#767]
### What are you trying to do?

perform the transformation A^T B A, where A is an arbitrary n x m matrix and B is a symmetric matrix (I am, perhaps mistakenly, calling this a shearing transformation), resulting in a mat.SymDense.

### What did you try?

looking at the GoDoc for the existence of such a method

### How does Gonum not allow you to achieve your goal?
I can do what I want to do if I am ok with a mat.Dense result, but I would prefer a SymDense since this transformation maintains the symmetric nature of the original matrix.

### What version of Go and Gonum are you using?

go 1.11, gonum v1 004553317c78aa27aeff99178a5a6bfe255b68a8

### Is this feature absent from the current master?
yes

### Are you able to help contribute the feature?

Hmmm, perhaps.  First I want feedback about whether or not I'm crazy for thinking this should be in there.

**Comment by kortschak** [comment]
ISTM that A is not a general m×n matrix, but rather a general n×n matrix to match the symmetric n×n of B.

Do you have a reference for use or a description of what you're trying to achieve (if this is possible)?

**Comment by decibelcooper** [comment]
No, I do mean that `A` is not square in general.  The transformation results in a new symmetric matrix that is in general a different size than `B`.

My particular case is in performing regression to covariant data.  In this case, `A` (n x m) is what is typically known as a "design matrix" that maps from a true parameter space (dimension m) to measurement space (dimension n): `y = A X + Delta`.  And `B` (n x n) is a symmetric pos. def. weighting matrix that comes into play in defining a `chi^2`.

Define a `chi^2` as the square of the Mahalanobis distance from the fit to the measurements, `chi^2 = delta^T B delta`, where `B` is the inverse of the expectation value of the outer product of the fit residuals `<delta delta^T>^{-1}`, and `delta \equiv y - A x` (note that `A` is symmetric pos. def., and `B^{-1}` is a covariance matrix).  Then, a solution for estimated parameters `x` can be found by minimizing `chi^2` WRT `x` which leads to `A^T B y = A^T B A x`.  This `A^T B A` matrix is symmetric and is what I would like to describe, perhaps with a `mat.SymDense`.

**Comment by kortschak** [comment]
Thanks for the explanation.

This should be easily computable with `mat.Inner` in a nested loop over the triangle of the symmetric matrix you are wanting to store it in. I don't see any simpler alternative, but maybe @btracey or @vladimir-ch know of one.

**Comment by decibelcooper** [comment]
Thank you for the suggestion.  I think that this is what I will do for now.

**Comment by btracey** [comment]
This is a sort of generalization of RankK / OuterK. It seems like a reasonable thing to add as a method. I wonder if there is a twist on RankK that would be a good name.

**Comment by deasmhumhna** [comment]
As a related idea, the `?gemmt` BLAS method for doing matrix multiplication that only returns the upper (or lower) triangle may be a useful intermediary.
 https://software.intel.com/en-us/mkl-developer-reference-fortran-gemmt
